### PR TITLE
Update Hostname resource to account for newer versions of PowerShell being standard

### PR DESCRIPTION
**Pull request by johnmccrae**
Updated the Windows code to use Rename-Computer instead of WMI, added properties for a domain username and password.

Signed-off-by: John McCrae <jmccrae@chef.io>

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Chore (non-breaking change that does not add functionality or fix an issue)

## Checklist:

- [ ] I have read the **CONTRIBUTING** document.
- [ ] I have run the pre-merge tests locally and they pass.
- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.
- [ ] All commits have been signed-off for [the Developer Certificate of Origin](https://github.com/chef/chef/blob/master/CONTRIBUTING.md#developer-certification-of-origin-dco).

### CODE REVIEW

File: lib/chef/resource/hostname.rb [hunk 1]
```diff
@@ -71,6 +89,12 @@ class Hostname < Chef::Resource
         description: "Determines whether or not Windows should be reboot after changing the hostname, as this is required for the change to take effect.",
         default: true
 
+      property :windows_domain_username, String,
```
Comment (tas50) [inline lib/chef/resource/hostname.rb:92]: Let's just call this `domain_user` and `domain_password` and set the description "Windows only: whatever whatever whatever"

Comment (tas50) [reply lib/chef/resource/hostname.rb:92]: Also domain_password should be sensitive: true so we don't leak that

File: lib/chef/resource/hostname.rb [hunk 2]
```diff
@@ -245,8 +269,15 @@ def updated_ec2_config_xml
...
-                $sysInfo = Get-WmiObject -Class Win32_ComputerSystem
-                $sysInfo.Rename("#{new_resource.hostname}")
+                if ([string]::IsNullOrEmpty(#{new_resource.windows_domain_username})){
+                  Rename-Computer -NewName #{new_resource.hostname}
+                }
+                else {
+                  $user = #{new_resource.windows_domain_username}
+                  $securepassword = #{new_resource.windows_user_password} | Convertto-SecureString -AsPlainText -Force
+                  $Credentials = New-Object System.Management.Automation.PSCredential -Argumentlist ($user, $securepassword)
+                  Rename-Computer -NewName #{new_resource.hostname} -DomainCredential $Credemtials
```
Comment (tas50) [inline lib/chef/resource/hostname.rb:279]: Credemtials? That's a typo right?

File: lib/chef/resource/hostname.rb [hunk 3]
```diff
@@ -245,8 +269,15 @@ def updated_ec2_config_xml
...
             converge_by "set hostname to #{new_resource.hostname}" do
               powershell_exec! <<~EOH
-                $sysInfo = Get-WmiObject -Class Win32_ComputerSystem
-                $sysInfo.Rename("#{new_resource.hostname}")
+                if ([string]::IsNullOrEmpty(#{new_resource.windows_domain_username})){
+                  Rename-Computer -NewName #{new_resource.hostname}
+                }
+                else {
+                  $user = #{new_resource.windows_domain_username}
+                  $securepassword = #{new_resource.windows_user_password} | Convertto-SecureString -AsPlainText -Force
```
Comment (tas50) [inline lib/chef/resource/hostname.rb:277]: If you make this secure_password you don't need to add it to the dictionary

File: cspell.json [hunk 4]
```diff
@@ -58,6 +58,7 @@
     "ARCHITEW",
     "archq",
     "argdup",
+    "Argumentlist",
```
Comment (tas50) [inline cspell.json:61]: I added Argumentlist to the global dictionary so this can get removed

File: cspell.json [hunk 5]
```diff
@@ -286,6 +286,7 @@
     "createhomedir",
     "Createobject",
     "creds",
+    "Credentials",
```
Comment (tas50) [inline cspell.json:289]: nuke this

**Review by lamont-granquist — APPROVED** [review]

**Pull request merged**